src/Npgsql/TypeHandlers/JsonbHandler.cs

```diff
@@ -107,6 +118,16 @@ public override int ValidateAndGetLength(char[] value, ref NpgsqlLengthCache len
             return await base.Read(buf, len - 1, async, fieldDescription);
         }
 
+        internal override async ValueTask<byte[]> ReadBytes(NpgsqlReadBuffer buf, int byteLen, bool async)
```

Comment:
If what you're looking for is efficiency, then you probably want to split this method into a non-async method calling a local `ReadBytesLong` function as seen elsewhere. In almost all cases the entire JSON document will be in memory (in fact, the entire row is always in memory unless `CommandBehavior.Sequential` is specified), and so we pay the overhead of the state machine for nothing.

Reply:
Yeah, we don't do this trick everywhere, only in places where perf is considered extra-important, since it causes some code duplication... But since the whole point of this PR is to do high-performance json decoding/encoding, it seems to make sense.